curp/src/server/storage/wal/pipeline.rs

```diff
@@ -0,0 +1,167 @@
+use std::{
+    io,
+    path::{Path, PathBuf},
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc,
+    },
+    task::Poll,
+};
+
+use clippy_utilities::OverflowArithmetic;
+use event_listener::Event;
+use flume::r#async::RecvStream;
+use futures::{FutureExt, StreamExt};
+use thiserror::Error;
+use tokio::task::JoinHandle;
+use tokio_stream::Stream;
+use tracing::error;
+
+use super::util::LockedFile;
+
+/// The temp file extension
+const TEMP_FILE_EXT: &str = ".tmp";
+
+/// The file pipeline, used for pipelining the creation of temp file
+pub(super) struct FilePipeline {
+    /// The directory where the temp files are created
+    dir: PathBuf,
+    /// The size of the temp file
+    file_size: u64,
+    /// The file receive stream
+    file_stream: RecvStream<'static, LockedFile>,
+    /// Stopped flag
+    stopped: Arc<AtomicBool>,
+}
+
+impl FilePipeline {
+    /// Creates a new `FilePipeline`
+    #[allow(clippy::integer_arithmetic)] // Introduced by tokio::select! macro
+    pub(super) fn new(dir: PathBuf, file_size: u64) -> io::Result<Self> {
+        Self::clean_up(&dir)?;
+
+        let (file_tx, file_rx) = flume::bounded(1);
+        let stop_event = Event::new();
+        let dir_c = dir.clone();
+        let stopped = Arc::new(AtomicBool::new(false));
+        let stopped_c = Arc::clone(&stopped);
+
+        let _ignore = tokio::spawn(async move {
+            let mut file_count = 0;
+            loop {
+                let file = match Self::alloc(&dir_c, file_size, file_count) {
```

Comment:
`Self::alloc` seems to be slow (preallocation) and it is blocking (i.e. no `.await`). Could we convert it into async io or spawn this task in blocking threads pool(`tokio::spawn_blocking`)?

Reply:
This may block the Tokio scheduler every time the log is sealed and generate the next log file.

Reply:
Tokio file does not seems to support preallocation, I refactored this to run in another thread.